src/irmin-pack/pack_store.ml

```diff
@@ -127,91 +181,211 @@ module Maker
       let buf = Bytes.create K.hash_size in
       let n = IO.read t.pack.block ~off buf in
       assert (n = K.hash_size);
-      decode_key (Bytes.unsafe_to_string buf) (ref 0)
+      decode_bin_hash (Bytes.unsafe_to_string buf) (ref 0)
+
+    let pack_file_contains_key t k =
+      let offset = Key.to_offset k and length = Key.to_length k in
+      if
+        Int63.compare
+          (Int63.add offset (Int63.of_int length))
+          (IO.offset t.pack.block)
+        > 0
+      then false
+      else
+        (* We read the hash explicitly as an integrity check: it's not
+           sufficient to assume that any offset within the file is valid. *)
+        let hash = io_read_and_decode_hash ~off:offset t in
+        let expected_hash = Key.to_hash k in
+        (* XXX: raise exception / log error in the [false] case here? *)
+        Hash.equal hash expected_hash
+
+    (* NOTE: may return false negatives, since the index may be partial. *)
+    (* let mem_via_hash t h =
+     *   Log.debug (fun l -> l "[pack] contains_hash %a" pp_hash h);
+     *   Tbl.mem t.staging h || Lru.mem t.lru h || Index.mem t.pack.index h *)
 
     let unsafe_mem t k =
-      [%log.debug "[pack] mem %a" pp_hash k];
-      Tbl.mem t.staging k || Lru.mem t.lru k || Index.mem t.pack.index k
+      [%log.debug "[pack] mem %a" pp_key k];
+      Tbl.mem t.staging (Key.to_hash k)
+      || Lru.mem t.lru (Key.to_hash k)
+      || pack_file_contains_key t k
 
     let mem t k =
       let b = unsafe_mem t k in
       Lwt.return b
 
-    let check_key k v =
-      let k' = Val.hash v in
-      if equal_key k k' then Ok () else Error (k, k')
+    let check_hash h v =
+      let h' = Val.hash v in
+      if equal_hash h h' then Ok () else Error (h, h')
 
-    exception Invalid_read
+    let check_key k v = check_hash (Key.to_hash k) v
 
     let io_read_and_decode ~off ~len t =
-      if (not (IO.readonly t.pack.block)) && off > IO.offset t.pack.block then
-        raise Invalid_read;
+      let () =
+        if not (IO.readonly t.pack.block) then
+          let io_offset = IO.offset t.pack.block in
+          if off > io_offset then
+            (* This is likely a store corruption. We raise [Invalid_read]
+               specifically so that [integrity_check] below can handle it. *)
+            invalid_read
+              "Got request to read %d bytes (at offset %a), but max IO offset \
+               is %a"
+              len Int63.pp off Int63.pp io_offset
+      in
       let buf = Bytes.create len in
       let n = IO.read t.pack.block ~off buf in
-      if n <> len then raise Invalid_read;
-      let hash off = io_read_and_decode_hash ~off t in
+      if n <> len then
+        invalid_read "Read %d bytes (at offset %a) but expected %d" n Int63.pp
+          off len;
+      let key_of_offset off =
+        (* XXX(craigfe): we may as well attempt to read the length here as well
+           (?), as we're touching this page anyway *)
```

Comment:
Agreed, done. This leads to an interesting current state:

When decoding new inodes, we eagerly fetch hash & length information of children. If the child is an old inode value, we find that there is no length header. At this point, there's no value in retaining the offset of that inode in memory: we can never decode it without first checking the index anyway. (In fact, with the current implementation it's strictly worse to retain the offset, because then on `find` we'll re-check the pack file and discover that there's no length header there.)

This means that there's no existing use-case for the `Direct_unknown_length` case of keys, except my WIP new commit value (but that one could just use `key_of_offset` to be symmetric with the inode case). We can probably remove it completely.